ambry-router/src/test/java/com.github.ambry.router/GetOperationTest.java

```diff
@@ -0,0 +1,474 @@
+    for (MockServer server : mockServerLayout.getMockServers()) {
+      if (i == indexToSetBlobDeleted) {
+        server.setServerErrorForAllRequests(ServerErrorCode.Blob_Deleted);
+      } else {
+        server.setServerErrorForAllRequests(ServerErrorCode.Blob_Not_Found);
+      }
+      i++;
+    }
+
+    while (!op.isOperationComplete()) {
+      op.poll(requestRegistrationCallback);
+      List<ResponseInfo> responses = sendAndWaitForResponses(requestListToFill);
+      for (ResponseInfo response : responses) {
+        op.handleResponse(response);
+        if (op.isOperationComplete()) {
+          break;
+        }
+      }
+    }
+
+    Assert.assertTrue("Operation should be complete at this time", op.isOperationComplete());
+    Assert.assertTrue(operationFuture.isDone());
+    RouterException routerException = (RouterException) operationException.get();
+    Assert.assertEquals(RouterErrorCode.BlobDeleted, routerException.getErrorCode());
+  }
+
+  /**
+   * Test the case with Blob_Not_Found errors from most servers, and Blob_Expired at just one server. The latter
+   * should be the exception received for the operation.
+   * @throws Exception
+   */
+  @Test
+  public void testErrorPrecedenceWithBlobExpiredCase()
+      throws Exception {
+    GetBlobInfoOperation op =
+        new GetBlobInfoOperation(routerConfig, mockClusterMap, responseHandler, blobIdStr, operationFuture,
+            operationCallback, time);
+    ArrayList<RequestInfo> requestListToFill = new ArrayList<>();
+    requestRegistrationCallback.requestListToFill = requestListToFill;
+
+    int i = 0;
+    int indexToSetBlobDeleted = random.nextInt(replicasCount);
+    for (MockServer server : mockServerLayout.getMockServers()) {
+      if (i == indexToSetBlobDeleted) {
+        server.setServerErrorForAllRequests(ServerErrorCode.Blob_Expired);
+      } else {
+        server.setServerErrorForAllRequests(ServerErrorCode.Blob_Not_Found);
+      }
+      i++;
+    }
+
+    while (!op.isOperationComplete()) {
+      op.poll(requestRegistrationCallback);
+      List<ResponseInfo> responses = sendAndWaitForResponses(requestListToFill);
+      for (ResponseInfo response : responses) {
+        op.handleResponse(response);
+        if (op.isOperationComplete()) {
+          break;
+        }
+      }
+    }
+
+    Assert.assertTrue("Operation should be complete at this time", op.isOperationComplete());
+    Assert.assertTrue(operationFuture.isDone());
+    RouterException routerException = (RouterException) operationException.get();
+    Assert.assertEquals(RouterErrorCode.BlobExpired, routerException.getErrorCode());
+  }
+
+  /**
+   * Test the case with multiple errors (server level and partition level) from multiple servers,
+   * with just one server returning a successful response. The operation should succeed.
+   * @throws Exception
+   */
+  @Test
+  public void testSuccessInThePresenceOfVariousErrors()
+      throws Exception {
+    GetBlobInfoOperation op =
```

Comment:
Probably. There are some subtleties to take care of when generalizing, such as here, where we need to ensure that one of the nodes in the local DC is set to No_Error. If we go down that path, we have to be careful to not end up complicating the tests when we try to handle such subtleties, and it might require some effort.